File: app/src/main/java/com/mehmetakiftutuncu/muezzin/fragments/preferences/PreferencesFragment.java
package com.mehmetakiftutuncu.muezzin.fragments.preferences;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.preference.Preference;
import android.preference.PreferenceFragment;

import com.mehmetakiftutuncu.muezzin.R;
import com.mehmetakiftutuncu.muezzin.activities.PlaceSelectionActivity;
import com.mehmetakiftutuncu.muezzin.activities.preferences.ReminderPreferencesActivity;
import com.mehmetakiftutuncu.muezzin.models.Place;
import com.mehmetakiftutuncu.muezzin.utilities.Pref;
import com.mehmetakiftutuncu.muezzin.utilities.optional.Optional;

/**
 * Created by akif on 08/05/16.
 */
public class PreferencesFragment extends PreferenceFragment {
    public static final String KEY_GENERAL_PLACE     = "general_place";
    public static final String KEY_GENERAL_REMINDERS = "general_reminders";
    public static final String KEY_MORE_RATE         = "more_rate";
    public static final String KEY_MORE_FEEDBACK     = "more_feedback";
    public static final String KEY_MORE_VERSION      = "more_version";
    public static final String KEY_MORE_LICENSES     = "more_licenses";

    public static final String RATE_URI = "market://details?id=com.mehmetakiftutuncu.muezzin";

    public static final String FEEDBACK_CONTACT = "m.akif.tutuncu@gmail.com";

    @Override public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        addPreferencesFromResource(R.xml.preferences);
    }

    @Override public void onResume() {
        super.onResume();

        initializePreferences();
    }

    private void initializePreferences() {
        Preference place = findPreference(KEY_GENERAL_PLACE);
        place.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override public boolean onPreferenceClick(Preference preference) {
                Intent intent = new Intent(getActivity(), PlaceSelectionActivity.class);
                intent.putExtra(PlaceSelectionActivity.EXTRA_STARTED_FROM_PREFERENCES, true);
                startActivity(intent);

                return true;
            }
        });

        updatePlaceSummary(place);

        Preference reminders = findPreference(KEY_GENERAL_REMINDERS);
        reminders.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override public boolean onPreferenceClick(Preference preference) {
                Intent intent = new Intent(getActivity(), ReminderPreferencesActivity.class);
                startActivity(intent);

                return true;
            }
        });

        Preference rate = findPreference(KEY_MORE_RATE);
        rate.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override public boolean onPreferenceClick(Preference preference) {
                Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(RATE_URI));
                startActivity(Intent.createChooser(intent, getString(R.string.preferences_more_rate)));

                return true;
            }
        });

        Preference feedback = findPreference(KEY_MORE_FEEDBACK);
        feedback.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override public boolean onPreferenceClick(Preference preference) {
                Intent intent = new Intent(Intent.ACTION_SEND);
                intent.setType("message/rfc822");
                intent.putExtra(Intent.EXTRA_EMAIL, new String[]{FEEDBACK_CONTACT});
                intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.applicationName));
                startActivity(Intent.createChooser(intent, getString(R.string.preferences_more_feedback)));

                return true;
            }
        });

        Preference version = findPreference(KEY_MORE_VERSION);
        String versionName = Pref.getAppVersionName(getActivity());
        version.setSummary(versionName);

        Preference licenses = findPreference(KEY_MORE_LICENSES);
    }

    private void updatePlaceSummary(Preference preference) {
        Optional<Place> maybeCurrentPlace = Pref.Places.getCurrentPlace(getActivity());

        String summary = maybeCurrentPlace.isDefined ? maybeCurrentPlace.get().getPlaceName(getActivity()).getOrElse("") : "";

        if (!summary.isEmpty()) {
            preference.setSummary(summary);
        }
    }
}

File: app/src/main/java/com/mehmetakiftutuncu/muezzin/fragments/preferences/PreferencesFragment.kt
package com.mehmetakiftutuncu.muezzin.fragments.preferences

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.mehmetakiftutuncu.muezzin.R
import com.mehmetakiftutuncu.muezzin.activities.LicencesActivity
import com.mehmetakiftutuncu.muezzin.activities.PlaceSelectionActivity
import com.mehmetakiftutuncu.muezzin.activities.preferences.ReminderPreferencesActivity
import com.mehmetakiftutuncu.muezzin.models.PrayerTimeReminder
import com.mehmetakiftutuncu.muezzin.repositories.PlaceRepository
import com.mehmetakiftutuncu.muezzin.utilities.Pref
import com.mehmetakiftutuncu.muezzin.widget.PrayerTimesWidget

class PreferencesFragment: PreferenceFragmentCompat() {
    private val ctx: Context by lazy { requireActivity() }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.preferences)
    }

    override fun onResume() {
        super.onResume()

        initializePreferences()

        PrayerTimeReminder.rescheduleReminders(ctx)
        PrayerTimesWidget.updateAllWidgets(ctx)
    }

    private fun initializePreferences() {
        findPreference<Preference>(keyGeneralPlace)?.apply {
            onPreferenceClickListener = Preference.OnPreferenceClickListener {
                Intent(ctx, PlaceSelectionActivity::class.java).apply {
                    putExtra(PlaceSelectionActivity.extraStartedFromPreferences, true)
                    startActivity(this)
                }
                true
            }

            updatePlaceSummary(this)
        }

        findPreference<Preference>(keyGeneralReminders)?.apply {
            onPreferenceClickListener = Preference.OnPreferenceClickListener {
                startActivity(Intent(ctx, ReminderPreferencesActivity::class.java))
                true
            }
        }

        findPreference<Preference>(keyMoreRate)?.apply {
            onPreferenceClickListener = Preference.OnPreferenceClickListener {
                Intent(Intent.ACTION_VIEW, Uri.parse(rateUri)).let {
                    Intent.createChooser(it, getString(R.string.preferences_more_rate))
                }.apply {
                    startActivity(this)
                }
                true
            }
        }

        findPreference<Preference>(keyMoreFeedback)?.apply {
            onPreferenceClickListener = Preference.OnPreferenceClickListener {
                Intent(Intent.ACTION_SEND).apply {
                    type = "message/rfc822"
                    putExtra(Intent.EXTRA_EMAIL, arrayOf(feedbackContact))
                    putExtra(Intent.EXTRA_SUBJECT, getString(R.string.applicationName))
                }.let {
                    Intent.createChooser(it, getString(R.string.preferences_more_feedback))
                }.apply {
                    startActivity(this)
                }
                true
            }

        }

        findPreference<Preference>(keyMoreVersion)?.apply {
            summary = Pref.appVersionName(ctx)
        }

        findPreference<Preference>(keyMoreLicenses)?.apply {
            onPreferenceClickListener = Preference.OnPreferenceClickListener {
                startActivity(Intent(ctx, LicencesActivity::class.java))
                true
            }
        }
    }

    private fun updatePlaceSummary(preference: Preference) =
        Pref.Places.getCurrentPlace(ctx)?.let { PlaceRepository.getName(ctx, it) }?.apply {
            preference.summary = this
        }

    companion object {
        const val keyGeneralPlace = "general_place"
        const val keyGeneralReminders = "general_reminders"
        const val keyMoreRate = "more_rate"
        const val keyMoreFeedback = "more_feedback"
        const val keyMoreVersion = "more_version"
        const val keyMoreLicenses = "more_licenses"

        const val rateUri = "market://details?id=com.mehmetakiftutuncu.muezzin"

        const val feedbackContact = "m.akif.tutuncu@gmail.com"
    }
}